psi4/src/psi4/lib3index/df_helper.cc

```diff
@@ -1551,282 +1747,303 @@ void DF_Helper::transform_core() {
+                    }
+                }
+            }
+            timer_off("DFH: (w|Qb)->(bw|Q)");
+            if(!MO_core_){
+                timer_on("DFH: MO to disk");
+                put_tensor(putf, Np, std::make_pair(0, bsize - 1), std::make_pair(0, wsize - 1),
+                           std::make_pair(begin, end), op);
+            }
 
-                        } else {  // (w|Qb)->(Q|wb)
+        // result is in Qpq format
+        } else if (std::get<2>(transf_[order_[ind]]) == 0) {  
 
-#pragma omp parallel for num_threads(nthreads_)
-                            for (size_t x = 0; x < block_size; x++) {
-                                for (size_t z = 0; z < wsize; z++) {
-#pragma omp simd
-                                    for (size_t y = 0; y < bsize; y++) {
-                                        Np[x * st1 + z * bsize + y] = Fp[z * st2 + x * bsize + y];
-                                    }
-                                }
-                            }
-                            timer_off("DFH: (w|Qb)->(bw|Q)");
-                            timer_on("DFH: MO to disk");
-                            put_tensor(putf, Np, std::make_pair(begin, end), std::make_pair(0, wsize - 1),
-                                       std::make_pair(0, bsize - 1), op);
-                        }
+            // (w|Qb)->(Q|bw)
+            #pragma omp parallel for num_threads(nthreads_)
+            for (size_t x = 0; x < rblock_size; x++) {
+                for (size_t z = 0; z < wsize; z++) {
+                    for (size_t y = 0; y < bsize; y++) {
+                        Np[(bcount + x) * bsize * wsize + y * wsize + z] 
+                            = Fp[z * bsize *rblock_size + x * bsize + y];
                     }
-                    timer_off("DFH: MO to disk");
                 }
             }
-        }
-    }
-    // outfile->Printf("\n     ==> DF_Helper:--End Transformations (disk)<==\n\n");
-
-    if (direct_) {
-        // total size allowed, in doubles
-        size_t total_mem =
-            (memory_ > wfinal * naux * 2 + naux_ * naux_ ? wfinal * naux : (memory_ - naux_ * naux_) / 2);
+            timer_off("DFH: (w|Qb)->(bw|Q)");
+            if(!MO_core_){
+                timer_on("DFH: MO to disk");
+                put_tensor(putf, Np, std::make_pair(begin, end), std::make_pair(0, bsize - 1),
+                           std::make_pair(0, wsize - 1), op);
+            }
+        
+        // result is in pQq format
+        } else {
+            
+            // (w|Qb)->(bQw)
+            #pragma omp parallel for num_threads(nthreads_)
+            for (size_t x = 0; x < rblock_size; x++) {
+                for (size_t y = 0; y < bsize; y++) {
+                    for (size_t z = 0; z < wsize; z++) {
+                        Np[y * lblock_size * wsize + (bcount + x) * wsize + z] 
+                            = Fp[z * bsize *rblock_size + x * bsize + y];
+                    }
+                }
+            }
```

Comment:
Ouch, I was worried about the aligned COPY and then I find this. Be curious to see how much time we spend on these operations as well.

Reply:
These operations are timed and included in `timer.dat` as `DFH: (w|Qb)->(bw|Q)`.  Although it might be preferable to switch to `DFH: Transpose (w|Qb)` after these latest edits.  I agree these are ugly operations, but they are necessary and fairly inconsequential in the wake of some of the O(N^4) operations occurring.

Reply:
Keep in mind that our N isn't all that large, in the DFJK example we noticed that we spent ~50-70% of our time in a strided N^3 copy while the N^4 GEMM that *should* have been the bottleneck.